bin/dist_get

```diff
@@ -0,0 +1,143 @@
+#!/bin/bash
+
+die() {
+	echo "$@" >&2
+	exit 1
+}
+
+have_binary() {
+	type "$1" > /dev/null
+}
+
+check_writeable() {
+	printf "" > "$1" && rm "$1"
+}
+
+download() {
+	local url="$1"
+	local output="$2"
+
+	if [ -z "$url" ] || [ -z "$output" ]; then
+		die "download takes exactly two arguments. was given '$@'"
+	fi
+
+	if ! check_writeable "$output"; then
+		die "download error: cannot write to $output"
+	fi
+
+	printf 'Downloading "%s" to "%s"\n' "$url" "$output"
+	if have_binary wget; then
+		wget "$url" -O "$output"
+	elif have_binary curl; then
+		curl --silent "$url" > "$output"
+	elif have_binary fetch; then
+		fetch "$url" -o "$output"
+	else
+		die "no binary found to download $url. exiting."
+	fi
+}
+
+unarchive() {
+	local archivetype="$1"
+	local infile="$2"
+	local outfile="$3"
+	local distname="$4"
+
+	if ! check_writeable "$outfile"; then
+		die "unarchive error: cannot write to $outfile"
+	fi
+
+	case $archivetype in
+		tar.gz)
+			if have_binary tar; then
+				echo "==> using 'tar' to extract binary from archive"
+				tar -x "$distname/$distname" -f "$infile" -O > "$outfile"
+			else
+				die "no binary on system for extracting tar files"
```

Comment:
It might be more explicit to give the name of the missing binary. For example:

```
				die "no 'tar' binary on system for extracting tar files"
```